File: mobile/screens/RegisterScreen.js
import { useState } from 'react';
import {
  View,
  Text,
  TextInput,
  Pressable,
  StyleSheet,
  ActivityIndicator,
  KeyboardAvoidingView,
  ScrollView,
  Platform,
} from 'react-native';
import { useTheme } from '../theme';
import { Logo, Eyebrow, Display } from '../components/Brand';
import PasswordRules, {
  isPasswordValid,
  MIN_PASSWORD_LENGTH,
} from '../components/PasswordRules';
import { register, saveToken } from '../api';

export default function RegisterScreen({ onRegistered, onGoToLogin }) {
  const { colors } = useTheme();

  const [form, setForm] = useState({
    firstName: '',
    lastName: '',
    email: '',
    username: '',
    password: '',
  });
  const [error, setError] = useState('');
  const [busy, setBusy] = useState(false);

  function set(field, value) {
    setForm((cur) => ({ ...cur, [field]: value }));
  }

  async function handleRegister() {
    setError('');

    if (!form.username.trim() || !form.password.trim() || !form.email.trim()) {
      setError('Username, email, and password are required.');
      return;
    }
    if (!form.email.includes('@')) {
      setError('Please enter a valid email address.');
      return;
    }
    if (!isPasswordValid(form.password)) {
      setError(`Password must be at least ${MIN_PASSWORD_LENGTH} characters.`);
      return;
    }

    setBusy(true);
    try {
      const data = await register({
        username: form.username.trim(),
        password: form.password,
        firstName: form.firstName.trim(),
        lastName: form.lastName.trim(),
        email: form.email.trim(),
      });
      await saveToken(data.token);
      onRegistered(data);
    } catch (e) {
      setError(e.message || 'Could not create account.');
    } finally {
      setBusy(false);
    }
  }

  const inputStyle = [
    styles.input,
    {
      backgroundColor: colors.surface,
      borderColor: colors.border,
      color: colors.text,
    },
  ];

  return (
    <KeyboardAvoidingView
      style={[styles.screen, { backgroundColor: colors.background }]}
      behavior={Platform.OS === 'ios' ? 'padding' : undefined}
    >
      <ScrollView contentContainerStyle={styles.inner} keyboardShouldPersistTaps="handled">
        <Logo />
        <Eyebrow style={{ marginTop: 26 }}>GET STARTED</Eyebrow>
        <Display size={30} style={{ marginTop: 10, marginBottom: 6 }}>
          Create your{'\n'}account.
        </Display>

        {error ? (
          <Text style={[styles.error, { color: colors.error }]}>{error}</Text>
        ) : null}

        <View style={styles.row}>
          <View style={styles.half}>
            <Text style={[styles.label, { color: colors.text }]}>First name</Text>
            <TextInput
              value={form.firstName}
              onChangeText={(v) => set('firstName', v)}
              placeholder="First"
              placeholderTextColor={colors.textMuted}
              style={inputStyle}
            />
          </View>
          <View style={styles.half}>
            <Text style={[styles.label, { color: colors.text }]}>Last name</Text>
            <TextInput
              value={form.lastName}
              onChangeText={(v) => set('lastName', v)}
              placeholder="Last"
              placeholderTextColor={colors.textMuted}
              style={inputStyle}
            />
          </View>
        </View>

        <Text style={[styles.label, { color: colors.text }]}>Email</Text>
        <TextInput
          value={form.email}
          onChangeText={(v) => set('email', v)}
          autoCapitalize="none"
          autoCorrect={false}
          keyboardType="email-address"
          placeholder="you@example.com"
          placeholderTextColor={colors.textMuted}
          style={inputStyle}
        />

        <Text style={[styles.label, { color: colors.text }]}>Username</Text>
        <TextInput
          value={form.username}
          onChangeText={(v) => set('username', v)}
          autoCapitalize="none"
          autoCorrect={false}
          placeholder="username"
          placeholderTextColor={colors.textMuted}
          style={inputStyle}
        />

        <Text style={[styles.label, { color: colors.text }]}>Password</Text>
        <TextInput
          value={form.password}
          onChangeText={(v) => set('password', v)}
          secureTextEntry
          autoCapitalize="none"
          placeholder="At least 6 characters"
          placeholderTextColor={colors.textMuted}
          style={inputStyle}
        />

        <PasswordRules password={form.password} />

        <Pressable
          onPress={handleRegister}
          disabled={busy}
          style={({ pressed }) => [
            styles.button,
            { backgroundColor: colors.primary, opacity: pressed || busy ? 0.7 : 1 },
          ]}
        >
          {busy ? (
            <ActivityIndicator color={colors.onPrimary} />
          ) : (
            <Text style={[styles.buttonText, { color: colors.onPrimary }]}>
              Create Account
            </Text>
          )}
        </Pressable>

        <Text style={[styles.note, { color: colors.textMuted }]}>
          We'll email you a verification link.
        </Text>

        <Pressable onPress={onGoToLogin} style={styles.linkWrap}>
          <Text style={[styles.link, { color: colors.textMuted }]}>
            Already have an account?{' '}
            <Text style={{ color: colors.primary }}>Log in</Text>
          </Text>
        </Pressable>
      </ScrollView>
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
  screen: { flex: 1 },
  inner: { padding: 28, paddingTop: 70, paddingBottom: 40 },
  brand: { fontSize: 30, fontWeight: '700', textAlign: 'center' },
  subtitle: { fontSize: 15, textAlign: 'center', marginTop: 6, marginBottom: 18 },
  row: { flexDirection: 'row', gap: 12 },
  half: { flex: 1 },
  label: { fontSize: 13, fontWeight: '600', marginBottom: 6, marginTop: 14 },
  input: {
    borderWidth: 1,
    borderRadius: 10,
    paddingHorizontal: 14,
    paddingVertical: 12,
    fontSize: 16,
  },
  button: {
    marginTop: 26,
    borderRadius: 10,
    paddingVertical: 15,
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: 50,
  },
  buttonText: { fontSize: 16, fontWeight: '700' },
  error: { fontSize: 14, textAlign: 'center', marginBottom: 4 },
  note: { fontSize: 12, textAlign: 'center', marginTop: 12 },
  linkWrap: { marginTop: 20, alignItems: 'center' },
  link: { fontSize: 14 },
});

File: noteriety_mobile/lib/screens/register_screen.dart
import 'package:flutter/material.dart';

import '../api/api.dart';
import '../theme/app_colors.dart';
import '../widgets/brand.dart';
import '../widgets/password_rules.dart';
import '../widgets/primary_button.dart';

class RegisterScreen extends StatefulWidget {
  final VoidCallback onRegistered; // token path (unused by this backend)
  final void Function(String email) onNeedsVerification; // -> code screen
  final VoidCallback onGoToLogin;

  const RegisterScreen({
    super.key,
    required this.onRegistered,
    required this.onNeedsVerification,
    required this.onGoToLogin,
  });

  @override
  State<RegisterScreen> createState() => _RegisterScreenState();
}

class _RegisterScreenState extends State<RegisterScreen> {
  final _first = TextEditingController();
  final _last = TextEditingController();
  final _email = TextEditingController();
  final _username = TextEditingController();
  final _password = TextEditingController();

  String _error = '';
  bool _busy = false;

  @override
  void dispose() {
    for (final c in [_first, _last, _email, _username, _password]) {
      c.dispose();
    }
    super.dispose();
  }

  Future<void> _handleRegister() async {
    setState(() => _error = '');
    final email = _email.text.trim();
    final username = _username.text.trim();
    final password = _password.text;

    if (username.isEmpty || password.trim().isEmpty || email.isEmpty) {
      setState(() => _error = 'Username, email, and password are required.');
      return;
    }
    if (!email.contains('@')) {
      setState(() => _error = 'Please enter a valid email address.');
      return;
    }
    if (!isPasswordValid(password)) {
      setState(() =>
          _error = 'Password must be at least $minPasswordLength characters.');
      return;
    }

    setState(() => _busy = true);
    try {
      final data = await register(
        username: username,
        password: password,
        firstName: _first.text.trim(),
        lastName: _last.text.trim(),
        email: email,
      );
      final token = data['token'];
      if (token is String && token.isNotEmpty) {
        await saveToken(token);
        if (!mounted) return;
        widget.onRegistered();
      } else {
        // Expected path: no token — the backend emailed a six-digit code and
        // staged a pending registration. Hand off to the verify screen.
        if (mounted) widget.onNeedsVerification(email);
      }
    } on ApiException catch (e) {
      if (mounted) setState(() => _error = e.message);
    } catch (_) {
      if (mounted) setState(() => _error = 'Could not create account.');
    } finally {
      if (mounted) setState(() => _busy = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;

    return Scaffold(
      backgroundColor: colors.background,
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.fromLTRB(28, 40, 28, 40),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const Logo(),
              const Eyebrow('GET STARTED', padding: EdgeInsets.only(top: 26)),
              const SizedBox(height: 10),
              const Display('Create your\naccount.', size: 30),
              const SizedBox(height: 6),
              if (_error.isNotEmpty)
                Padding(
                  padding: const EdgeInsets.only(top: 6),
                  child: Text(
                    _error,
                    textAlign: TextAlign.center,
                    style: TextStyle(color: colors.error, fontSize: 14),
                  ),
                ),
              Row(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        _label('First name'),
                        AppTextField(controller: _first, hint: 'First'),
                      ],
                    ),
                  ),
                  const SizedBox(width: 12),
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        _label('Last name'),
                        AppTextField(controller: _last, hint: 'Last'),
                      ],
                    ),
                  ),
                ],
              ),
              _label('Email'),
              AppTextField(
                controller: _email,
                hint: 'you@example.com',
                keyboardType: TextInputType.emailAddress,
                capitalization: TextCapitalization.none,
              ),
              _label('Username'),
              AppTextField(
                controller: _username,
                hint: 'username',
                capitalization: TextCapitalization.none,
              ),
              _label('Password'),
              AppTextField(
                controller: _password,
                hint: 'At least 8 characters',
                obscure: true,
                capitalization: TextCapitalization.none,
              ),
              // Live rules react to typing.
              ValueListenableBuilder(
                valueListenable: _password,
                builder: (_, value, __) =>
                    PasswordRules(password: value.text),
              ),
              PrimaryButton(
                label: 'Create Account',
                busy: _busy,
                onPressed: _handleRegister,
                margin: const EdgeInsets.only(top: 26),
              ),
              Padding(
                padding: const EdgeInsets.only(top: 12),
                child: Center(
                  child: Text(
                    "We'll email you a six-digit code.",
                    style: TextStyle(fontSize: 12, color: colors.textMuted),
                  ),
                ),
              ),
              const SizedBox(height: 8),
              Center(
                child: GestureDetector(
                  onTap: widget.onGoToLogin,
                  child: Text.rich(
                    TextSpan(
                      text: 'Already have an account? ',
                      style: TextStyle(color: colors.textMuted, fontSize: 14),
                      children: [
                        TextSpan(
                          text: 'Log in',
                          style: TextStyle(color: colors.primary),
                        ),
                      ],
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _label(String text) => Padding(
        padding: const EdgeInsets.only(top: 14, bottom: 6),
        child: Text(
          text,
          style: TextStyle(
            fontSize: 13,
            fontWeight: FontWeight.w600,
            color: context.colors.text,
          ),
        ),
      );
}